pkg/role/sdkserviceapi.go

```
@@ -0,0 +1,357 @@
+func matchRule(rule, s string) bool {
+	// no rule
+	rl := len(rule)
+	if rl == 0 {
+		return false
+	}
+
+	// '*xxx' || 'xxx*'
+	if rule[0:1] == "*" || rule[rl-1:rl] == "*" {
+		// get the matching string from the rule
+		match := strings.TrimSpace(strings.Join(strings.Split(rule, "*"), ""))
+
+		// '*' or '*******'
+		if len(match) == 0 {
+			return true
+		}
+
+		// '*xxx*'
+		if rule[0:1] == "*" && rule[rl-1:rl] == "*" {
+			return strings.Contains(s, match)
+		}
+
+		// '*xxx'
+		if rule[0:1] == "*" {
+			return strings.HasSuffix(s, match)
+		}
+
+		// 'xxx*'
+		return strings.HasPrefix(s, match)
+	}
+
+	// no wildcard stars given in rule
+	return rule == s
+}
+
+// NewSdkSErviceAPIRoles returns a new SDK role manager
+func NewSdkRoleManager(kv kvdb.Kvdb) (*SdkRoleManager, error) {
+	s := &SdkRoleManager{
+		kv: kv,
+	}
+
+	// Load all default roles
+	for k, v := range defaultRoles {
+		role := &api.SdkRole{
+			Name:  k,
+			Rules: v,
+		}
+		if _, err := kv.Put(prefixWithName(k), role, 0); err != nil {
+			return nil, err
+		}
+	}
+
+	return s, nil
+}
+
+// Create saves a role in Kvdb
+func (r *SdkRoleManager) Create(
+	ctx context.Context,
+	req *api.SdkRoleCreateRequest,
+) (*api.SdkRoleCreateResponse, error) {
+	if req.GetRole() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Must supply a role")
+	}
+	if err := r.checkRoleForErrors(req.GetRole()); err != nil {
+		return nil, err
+	}
+
+	// Determine if there is collision with default roles
+	if _, ok := defaultRoles[req.GetRole().GetName()]; ok {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"Name %s already used by system role", req.GetRole().GetName())
+	}
+
+	// Reviewer:
+	// Should we check here if the value already exists and return the object
+	// currently in the db? That is is more accurate and idempotent.
+	// This will require a change to the return value of this function (and Update)
+
+	_, err := r.kv.Put(prefixWithName(req.GetRole().GetName()), req.GetRole(), 0)
```

You should use the API kv.Create(). If the key already exists it will return an error.
The intent of the user (and the API) was to create, if there exist a role we should return an appropriate error. If we do not return an error and return the existing role, then it is confusing.